### ISSUE TITLE
Malformed JavaScript resource URLs in zipped reports

### ISSUE THREAD

**Issue by hannes-brt** [snakemake/snakemake#862]
**Snakemake version**
5.32.0

Tested with MacOS 11.1, conda 4.9.2, Safari 14.0.2, Google Chrome 88.0.4324.96

**Describe the bug**
When creating and opening a zipped report, any pages under "Results" show "Loading..." forever and never display the added materials.
This seems to be due to malformed JavaScript resource URLs, e.g. `file://null/<path_to_report>/feather.min.js.map`. These URLs are blocked by most browsers sandbox, but they are also malformed and don't exist in the first place.

**Logs**
```
[Error] SyntaxError: Unexpected token ','
	(anonymous function) (report.html:921)
[Error] Not allowed to load local resource: file:///favicon.ico
[Error] Not allowed to load local resource: file:///extensions/MathMenu.js?V=2.7.4
	JS (report.html:593:14130)
	Load (report.html:593:13073)
	Require (report.html:593:12761)
	execute (report.html:593:3248)
	cb (report.html:593:2825)
	Process (report.html:593:8712)
	Push (report.html:593:8512)
	Init (report.html:593:8240)
	Init (report.html:593:1036)
	(anonymous function) (report.html:593:593)
	(anonymous function) (report.html:593:49620)
[Error] Not allowed to load local resource: file:///extensions/MathZoom.js?V=2.7.4
	JS (report.html:593:14130)
	Load (report.html:593:13073)
	Require (report.html:593:12761)
	execute (report.html:593:3248)
	cb (report.html:593:2825)
[Error] TypeError: undefined is not an object (evaluating 'b[a]')
	(anonymous function) (report.html:657:93)
	R (report.html:654:394)
	e (report.html:730:471)
	(anonymous function) (report.html:731:200)
	each (report.html:560:2578)
	q (report.html:721:393)
	(anonymous function) (report.html:804:498)
	(anonymous function) (report.html:981)
	each (report.html:560:2578)
	show_category (report.html:970)
	show_panel (report.html:1057)
	(anonymous function) (report.html:1232)
	dispatch (report.html:560:41778)
[Error] Not allowed to load local resource: file://null/Users/hannes/Desktop/snakemake_js_bug/report/feather.min.js.map
[Error] Not allowed to request resource
[Error] Cannot load file://null/Users/hannes/Desktop/snakemake_js_bug/report/feather.min.js.map due to access control checks.
[Error] Not allowed to load local resource: file://null/Users/hannes/Desktop/snakemake_js_bug/report/bootstrap.min.js.map
[Error] Not allowed to request resource
[Error] Cannot load file://null/Users/hannes/Desktop/snakemake_js_bug/report/bootstrap.min.js.map due to access control checks.
[Error] Not allowed to load local resource: file://null/Users/hannes/Desktop/snakemake_js_bug/report/popper.min.js.map
[Error] Not allowed to request resource
[Error] Cannot load file://null/Users/hannes/Desktop/snakemake_js_bug/report/popper.min.js.map due to access control checks.
[Error] Not allowed to load local resource: file://null/Users/hannes/Desktop/snakemake_js_bug/report/ekko-lightbox.min.js.map
[Error] Not allowed to request resource
[Error] Cannot load file://null/Users/hannes/Desktop/snakemake_js_bug/report/ekko-lightbox.min.js.map due to access control checks.
```

**Minimal example**
Try the following Snakefile:
```
rule report:
  output:
    report("foo.txt", category="Foo")
  shell:
    "echo 'bar' > {output}"
```

```
$ snakemake -j 1 report
$ snakemake -j 1 --report report.zip
```

Then unzip `report.zip`, open `report.html` and try to navigate to "Results | Foo".

**Comment by johanneskoester** [comment]
@FelixMoelder can you please try to reproduce this with your mac?

**Comment by hannes-brt** [comment]
@johanneskoester @FelixMoelder I might be wrong of course, but this doesn't seem to be a Mac specific bug to me.

**Comment by johanneskoester** [comment]
We use and generate the report all the time on linux, and this never happened before.

**Comment by hannes-brt** [comment]
I just generated the above example report on a Linux server and downloaded it. I still see the same issue. So if this is Mac specific, it's a problem with rendering the report rather than generating it.

To compare, I just downloaded https://github.com/snakemake/snakemake/blob/master/tests/test_report_zip/expected-results/report.zip (generated with Snakemake 5.13.0+34.g0a473b6d.dirty) locally, opened it on my Mac and confirmed it works fine on my Mac in Safari. However, I still see the same errors about loading JavaScript resources in the console, so this may suggest that the issue of the report not loading is separate from loading those JS resources.

**Comment by johanneskoester** [comment]
@hannes-brt can you try adding crossorigin to the script tags that fail to load (see [here](https://stackoverflow.com/a/60475081/7070491)?

**Comment by hannes-brt** [comment]
I'm not sure how Snakemake report does it, but I can't find any `<script>` tags that load external resources, only inlined JS. I'll append one of the problematic reports, maybe you can see what's going on.
[report.zip](https://github.com/snakemake/snakemake/files/5915422/report.zip)

**Comment by FelixMoelder** [comment]
I've just tried the minimal example on macOS and Linux and opened both on macOS 11.2 in Safari and Firefox.
Both reports open up with the same syntax error @hannes-brt reported.
So this does not appear to me as a Mac specific bug.

What's confusing me is, that I created a report of an internal pipeline this morning and that one does not show any issue (using snakemake 0.32.0 on Linux).

Moreover I noticed that the malformed resource path only shows up in Safari (in all three reports). This does not occur in Firefox which is why I would expect that as an issue caused by Safari.
Nevertheless, the malformed paths do not impact the functionality as the resources do not even exist.
It looks more like Safari is trying to pulling files being part of an inline command e.g. `//#sourceMappingURL=feather.min.js.map</script>`

**Comment by hannes-brt** [comment]
I'm thinking that I might have been wrong in attributing the problem to the JS resource load errors. As I said before if I download a zipped report from the Snakemake unit tests, it works fine for display even though it also shows the resource load errors.
The real problem is that the result pages hang forever on "Loading...". @FelixMoelder have you been able to confirm that problem?

**Comment by FelixMoelder** [comment]
Yes, that issue exists when opening the report created by your example and it looks like that's connected to the syntax error you reported.

The affected part of the code looks like this
```
var results = {
      "foo.txt": {
        name: "foo.txt",
        path: "foo.txt",
        size: "4 Bytes",
        caption: ,
        job_properties: {
          rule: "report",
          wildcards: "",
          params: ""
        },
        data_uri: function() { return "data/raw/ddab29ff2c393ee52855d21a240eb05f775df88e3ce347df759f0c4b80356c35/foo.txt"; },
        thumbnail_uri: function() { return null; }
      }
```

The attribute `caption` needs some value or empty string.
If you replace this line in `report.html` by `caption: "",` the loading issue is solved for me.
So it seems like there is a bug during creation of the report.

**Comment by hannes-brt** [comment]
That looks like the culprit: I can confirm that when I give a caption in the `report()` output, loading works fine.

**Comment by FelixMoelder** [comment]
https://github.com/snakemake/snakemake/pull/863 should fix this.

**Comment by johanneskoester** [comment]
PR is merged! Thanks Felix and Hannes!

**Comment by hannes-brt** [comment]
Hi Felix and Johannes,

I just had a chance to test the new 5.32.1 release including the fix and it seems that it doesn't yet solve the issue. What I found is that when a caption is given, it is now double-quoted such which also results in the result page getting stuck on "Loading....".

```
   caption: ""<p>Caption from rst file.</p>""
```

To confirm, I replaced all double quotes by single ones and that resolved the loading issue.

**Comment by FelixMoelder** [comment]
Hi @hannes-brt,

you are right. My PR just moved the issue as the caption already contains quotes.
I just created a new PR setting `""` as default in case no caption is set.

This should hopefully solve the issue. At least during testing both cases work as expected.
